processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java

```diff
@@ -889,15 +900,34 @@ public Metadata getMetadata()
       {
         final List<DimensionDesc> dimensions = getDimensions();
 
-        Map<TimeAndDims, Integer> facts = null;
-        if (descending && sortFacts) {
-          facts = ((ConcurrentNavigableMap<TimeAndDims, Integer>) getFacts()).descendingMap();
+        final Iterable<Map.Entry<TimeAndDims, Integer>> facts;
+        if (sorted && sortFacts) {
+          facts = descending ? ((ConcurrentNavigableMap<TimeAndDims, Integer>) getFacts()).descendingMap().entrySet()
+                             : getFacts().entrySet();
+        } else if (sorted) {
+          // Materialize and sort
+          final Comparator<TimeAndDims> comparator = descending
+                                                     ? Ordering.from(dimsComparator()).reverse()
+                                                     : dimsComparator();
+          List<Map.Entry<TimeAndDims, Integer>> factsList = Lists.newArrayList(getFacts().entrySet());
```

Comment:
I was thinking that for the current offheap impl this is fine (even the offheap one still has an onheap facts map) but things are different if the facts map is offheap.

I am ok waiting on this PR until #2847 is settled and then figuring out what is best.